Sources/Bond/AppKit/NSTableView.swift

```diff
@@ -196,7 +196,7 @@ public extension SignalProtocol where
       property: dataSource,
       to: #selector(NSTableViewDataSource.tableView(_:objectValueFor:row:)),
       map: { (dataSource: DataSource?, _: NSTableView, _: NSTableColumn, row: Int) -> Any? in
-        return dataSource?[row]
+        return dataSource?[row] as AnyObject
```

Comment:
@tonyarnold would you know how this affects things?

Reply:
To be honest, I'm not sure what the intent here is. The closure still returns `Any`, and it's not a conditional downcast… I don't think it's necessary?

@nayzak do you recall why you needed this?

Reply:
Hi @srdanrasic, @tonyarnold! Did you look at my demo that reproduces the bug? I've just updated it to Swift 4 and latest Bond release an it still crashes without casting to `AnyObject`.

Reply:
Just had a look now - wow, that's… special.

I'm honestly not sure where this issue is coming from - I'm not saying we shouldn't fix it, but I'd like to know what knock-on effects we're going to see from casting absolutely everything to `AnyObject` (which is essentially saying "this is a reference type now", right?)

Reply:
It was surprising for me too :) But obvious thing is that something goes wrong on a "glue level between Swift and ObjC". I've just made an assumption that objc definitely needs object, however Apple made `Any` same thing as `id`. Anyway I'm not that guy who knows what's actually going on in this case.

Reply:
I've been playing with this a bit and apparently we need to cast it in order for the compiler to correctly bridge value types to ObjC types: String -> NSString, Int -> NSNumber, etc. 

I will see if this can be fixed in `ProtocolProxy` class implementation where the bridging is supposed to happen but  does not, probably due to actual type is hidden under generic.

Reply:
Fixed in 6.4.2!

I had to manually bridge types, but it was worth it as it also improves other kinds of delegates.

Reply:
You're a goddamned wizard, 'arry. 🌟

Reply:
Far from that 😅

Reply:
Oh dagnabbit, I updated to 6.2.4 today, and now whatever changes occurred here are causing _my_ code to crash. 

<details>
```
(lldb) bt
* thread #1, queue = 'com.apple.main-thread', stop reason = EXC_BAD_ACCESS (code=EXC_I386_GPFLT)
    frame #0: 0x00007fff7ec19181 libobjc.A.dylib`objc_release + 33
    frame #1: 0x00000001039b9797 libswiftCore.dylib`swift::metadataimpl::ValueWitnesses<swift::metadataimpl::ObjCRetainableBox>::assignWithTake(swift::OpaqueValue*, swift::OpaqueValue*, swift::TargetMetadata<swift::InProcess> const*) + 23
    frame #2: 0x000000010396d2e0 libswiftCore.dylib`assignWithTake value witness for Swift.ImplicitlyUnwrappedOptional + 192
  * frame #3: 0x0000000102849af6 Bond`write #1 <A><A1, B1>(value=<unavailable>, type=<unavailable>, self=0x000060300008d370) in BNDInvocation.writeReturnValue<A>(_:) at ProtocolProxy.swift:78
    frame #4: 0x0000000102848cdd Bond`BNDInvocation.writeReturnValue<A>(value=<unavailable>, self=0x000060300008d370) at ProtocolProxy.swift:101
    frame #5: 0x00000001028526e0 Bond`closure #1 in ProtocolProxy.registerInvoker3<A, B, C, D>(invocation=0x000060300008d370, block=0x0000000102870160 Bond`partial apply forwarder for closure #1 (A, B, C) -> E in closure #1 (ReactiveKit.PublishSubject<D, ReactiveKit.NoError>) -> ReactiveKit.Disposable in Bond.ProtocolProxy.signal<A, B, C, D, E>(for: ObjectiveC.Selector, dispatch: (ReactiveKit.PublishSubject<D, ReactiveKit.NoError>, A, B, C) -> E) -> ReactiveKit.Signal<D, ReactiveKit.NoError> at ProtocolProxy.swift) at ProtocolProxy.swift:179
    frame #6: 0x000000010286e258 Bond`partial apply for closure #1 in ProtocolProxy.registerInvoker3<A, B, C, D>(for:block:) at ProtocolProxy.swift:0
    frame #7: 0x000000010284c419 Bond`thunk for @callee_owned (@owned BNDInvocation) -> () at ProtocolProxy.swift:0
    frame #8: 0x000000010286e2ea Bond`thunk for @callee_owned (@owned BNDInvocation) -> ()partial apply at ProtocolProxy.swift:0
    frame #9: 0x000000010284c37f Bond`thunk for @callee_owned (@in BNDInvocation) -> (@out ()) at ProtocolProxy.swift:0
    frame #10: 0x00000001028628d2 Bond`thunk for @callee_owned (@in BNDInvocation) -> (@out ())partial apply at ProtocolProxy.swift:0
    frame #11: 0x000000010284c983 Bond`ProtocolProxy.handle(invocation=0x000060300008d370, self=0x0000606000361160) at ProtocolProxy.swift:137
    frame #12: 0x000000010284caeb Bond`@objc ProtocolProxy.handle(_:) at ProtocolProxy.swift:0
    frame #13: 0x00000001028072fb Bond`-[BNDProtocolProxyBase forwardInvocation:](self=0x0000606000361160, _cmd="forwardInvocation:", invocation=0x00006060000ae8c0) at BNDProtocolProxyBase.m:86
    frame #14: 0x00007fff582ac24c CoreFoundation`___forwarding___ + 748
    frame #15: 0x00007fff582abed8 CoreFoundation`__forwarding_prep_0___ + 120
    frame #16: 0x00007fff55a5a41f AppKit`-[NSTableView(NSTableViewViewBased) _delegate_viewForTableColumn:row:] + 86
    frame #17: 0x00007fff559d4b90 AppKit`-[NSTableView(NSTableViewViewBased) makeViewForTableColumn:row:] + 87
    frame #18: 0x00007fff559d4413 AppKit`-[NSTableRowData _addViewToRowView:atColumn:row:] + 342
    frame #19: 0x00007fff559d40eb AppKit`-[NSTableRowData _addViewsToRowView:atRow:] + 211
    frame #20: 0x00007fff559d276a AppKit`-[NSTableRowData _initializeRowView:atRow:] + 397
    frame #21: 0x00007fff562bd1a8 AppKit`-[NSTableRowData _preparedRowViewForRow:storageHandler:] + 154
    frame #22: 0x00007fff559d0b97 AppKit`-[NSTableRowData _addRowViewForVisibleRow:withPriorView:] + 392
    frame #23: 0x00007fff562ba3de AppKit`-[NSTableRowData _slideInsertRow:intoClipView:withPriorView:] + 45
    frame #24: 0x00007fff55a942c8 AppKit`-[NSTableRowData _updateVisibleViewsBasedOnUpdateItems] + 3976
    frame #25: 0x00007fff55a931f6 AppKit`-[NSTableRowData _updateVisibleViewsBasedOnUpdateItemsAnimated] + 234
    frame #26: 0x00007fff559c3f54 AppKit`-[NSTableRowData _doWorkAfterEndUpdates] + 85
    frame #27: 0x00007fff559c3e00 AppKit`-[NSTableView _endUpdateWithTile:] + 137
    frame #28: 0x00000001028339e9 Bond`DefaultTableViewBond.apply(event=<unavailable>, tableView=0x000061400001d640, self=0x00006080001c2ba0) at NSTableView.swift:128
    frame #29: 0x00000001028355aa Bond`protocol witness for TableViewBond.apply(event:to:) in conformance <A> DefaultTableViewBond<A> at NSTableView.swift:0
    frame #30: 0x000000010283cb74 Bond`closure #5 in SignalProtocol<A>._bind<A>(tableView=0x000061400001d640, event=0x00007ffeefbf3ce0, dataSource=<unavailable>, bond=<unavailable>) at NSTableView.swift:206
    frame #31: 0x00000001028433db Bond`partial apply for closure #5 in SignalProtocol<A>._bind<A>(to:using:) at NSTableView.swift:0
    frame #32: 0x0000000102c4fa9a ReactiveKit`closure #1 in closure #1 in SignalProtocol<A>.bind<A>(target=<unavailable>, setter=0x0000000102843320 Bond`partial apply forwarder for closure #5 (__ObjC.NSTableView, A.Element) -> () in (extension in Bond):ReactiveKit.SignalProtocol<A where A: ReactiveKit.SignalProtocol, A.Element: Bond.DataSourceEventProtocol, A.Error == ReactiveKit.NoError, A.Element.DataSource: Bond.QueryableDataSourceProtocol, A.Element.BatchKind == Bond.BatchKindPatch, A.Element.DataSource.Index == Swift.Int>.(_bind in _BD200F2498C7964CF80293E5276838A7)<A where A1: Bond.TableViewBond, A1.DataSource: Bond.QueryableDataSourceProtocol, A1.DataSource == A.Element.DataSource, A1.DataSource.Index == Swift.Int>(to: __ObjC.NSTableView, using: A1) -> ReactiveKit.Disposable at NSTableView.swift, element=0x00006040002735a0) at Bindable.swift:110
    frame #33: 0x0000000102c513d9 ReactiveKit`partial apply for closure #1 in closure #1 in SignalProtocol<A>.bind<A>(to:context:setter:) at Bindable.swift:0
    frame #34: 0x0000000102c6e205 ReactiveKit`closure #1 in static ExecutionContext.immediateOnMain.getter(block=0x0000000102c51330 ReactiveKit`partial apply forwarder for closure #1 () -> () in closure #1 (A.Element) -> () in (extension in ReactiveKit):ReactiveKit.SignalProtocol<A where A: ReactiveKit.SignalProtocol, A.Error == ReactiveKit.NoError>.bind<A where A1: ReactiveKit.Deallocatable>(to: A1, context: ReactiveKit.ExecutionContext, setter: (A1, A.Element) -> ()) -> ReactiveKit.Disposable at Bindable.swift) at ExecutionContext.swift:60
    frame #35: 0x0000000102c6dc82 ReactiveKit`ExecutionContext.execute(block=0x0000000102c51330 ReactiveKit`partial apply forwarder for closure #1 () -> () in closure #1 (A.Element) -> () in (extension in ReactiveKit):ReactiveKit.SignalProtocol<A where A: ReactiveKit.SignalProtocol, A.Error == ReactiveKit.NoError>.bind<A where A1: ReactiveKit.Deallocatable>(to: A1, context: ReactiveKit.ExecutionContext, setter: (A1, A.Element) -> ()) -> ReactiveKit.Disposable at Bindable.swift, self=(context = 0x0000000102c6df40 ReactiveKit`closure #1 (() -> ()) -> () in static ReactiveKit.ExecutionContext.immediateOnMain.getter : ReactiveKit.ExecutionContext at ExecutionContext.swift:58)) at ExecutionContext.swift:47
    frame #36: 0x0000000102c4f41d ReactiveKit`closure #1 in SignalProtocol<A>.bind<A>(element=0x00007ffeefbf43a0, context=(context = 0x0000000102c6df40 ReactiveKit`closure #1 (() -> ()) -> () in static ReactiveKit.ExecutionContext.immediateOnMain.getter : ReactiveKit.ExecutionContext at ExecutionContext.swift:58), target=<unavailable>, setter=0x0000000102843320 Bond`partial apply forwarder for closure #5 (__ObjC.NSTableView, A.Element) -> () in (extension in Bond):ReactiveKit.SignalProtocol<A where A: ReactiveKit.SignalProtocol, A.Element: Bond.DataSourceEventProtocol, A.Error == ReactiveKit.NoError, A.Element.DataSource: Bond.QueryableDataSourceProtocol, A.Element.BatchKind == Bond.BatchKindPatch, A.Element.DataSource.Index == Swift.Int>.(_bind in _BD200F2498C7964CF80293E5276838A7)<A where A1: Bond.TableViewBond, A1.DataSource: Bond.QueryableDataSourceProtocol, A1.DataSource == A.Element.DataSource, A1.DataSource.Index == Swift.Int>(to: __ObjC.NSTableView, using: A1) -> ReactiveKit.Disposable at NSTableView.swift) at Bindable.swift:112
    frame #37: 0x0000000102c4f61b ReactiveKit`partial apply for closure #1 in SignalProtocol<A>.bind<A>(to:context:setter:) at Bindable.swift:0
    frame #38: 0x0000000102b93c7c ReactiveKit`closure #1 in SignalProtocol.observeNext(event=<unavailable>, observer=0x0000000102c4f550 ReactiveKit`partial apply forwarder for closure #1 (A.Element) -> () in (extension in ReactiveKit):ReactiveKit.SignalProtocol<A where A: ReactiveKit.SignalProtocol, A.Error == ReactiveKit.NoError>.bind<A where A1: ReactiveKit.Deallocatable>(to: A1, context: ReactiveKit.ExecutionContext, setter: (A1, A.Element) -> ()) -> ReactiveKit.Disposable at Bindable.swift) at SignalProtocol.swift:57
    frame #39: 0x0000000102b93ece ReactiveKit`partial apply for closure #1 in SignalProtocol.observeNext(with:) at SignalProtocol.swift:0
    frame #40: 0x0000000102c49c44 ReactiveKit`AtomicObserver.on(event=<unavailable>, self=0x000060c0000bf800) at Observer.swift:83
    frame #41: 0x0000000102c4adf2 ReactiveKit`protocol witness for ObserverProtocol.on(_:) in conformance <A, B> AtomicObserver<A, B> at Observer.swift:0
    frame #42: 0x0000000102c4b187 ReactiveKit`ObserverProtocol.next(element=0x00007ffeefbf4ec0, self=0x000000010ab9c2a0) at Observer.swift:97
    frame #43: 0x0000000102bd35d1 ReactiveKit`closure #2 in closure #1 in SignalProtocol.take<A>(event=<unavailable>, observer=0x000060c0000bf800) at SignalProtocol.swift:758
    frame #44: 0x0000000102c3f9ae ReactiveKit`partial apply for closure #2 in closure #1 in SignalProtocol.take<A>(until:) at SignalProtocol.swift:0
    frame #45: 0x0000000102c49c44 ReactiveKit`AtomicObserver.on(event=<unavailable>, self=0x000060c00009e080) at Observer.swift:83
    frame #46: 0x0000000102c4adf2 ReactiveKit`protocol witness for ObserverProtocol.on(_:) in conformance <A, B> AtomicObserver<A, B> at Observer.swift:0
    frame #47: 0x0000000102c4b187 ReactiveKit`ObserverProtocol.next(element=0x00007ffeefbf69f0, self=0x00007ffeefbf69d0) at Observer.swift:97
    frame #48: 0x00000001028c8d1d Bond`closure #1 in closure #1 in SignalProtocol<A>.toPatches(event=<unavailable>, isBatching=0x000060300043c5e0, originalArray=<unavailable>, observer=<unavailable>) at ObservableArray.swift:790
    frame #49: 0x00000001028c9754 Bond`partial apply for closure #1 in closure #1 in SignalProtocol<A>.toPatches() at ObservableArray.swift:0
    frame #50: 0x000000010289458d Bond`thunk for @callee_owned (@in Event<ObservableArrayEvent<A>, NoError>) -> () at ObservableArray.swift:0
    frame #51: 0x00000001028c5962 Bond`partial apply for thunk for @callee_owned (@in Event<ObservableArrayEvent<A>, NoError>) -> () at ObservableArray.swift:0
    frame #52: 0x0000000102893d26 Bond`thunk for @callee_owned (@owned Event<ObservableArrayEvent<A>, NoError>) -> () at ObservableArray.swift:0
    frame #53: 0x0000000102893e22 Bond`partial apply for thunk for @callee_owned (@owned Event<ObservableArrayEvent<A>, NoError>) -> () at ObservableArray.swift:0
    frame #54: 0x0000000102b4aba7 ReactiveKit`thunk for @callee_owned (@in Event<A, B>) -> () at Subjects.swift:0
    frame #55: 0x0000000102b5b0a6 ReactiveKit`partial apply for thunk for @callee_owned (@in Event<A, B>) -> () at Subjects.swift:0
    frame #56: 0x0000000102b4bda7 ReactiveKit`thunk for @callee_owned (@in Event<A, B>) -> (@out ()) at Subjects.swift:0
    frame #57: 0x0000000102b5b5ce ReactiveKit`thunk for @callee_owned (@in Event<A, B>) -> (@out ())partial apply at Subjects.swift:0
    frame #58: 0x0000000102b49548 ReactiveKit`closure #1 in Subject.send($0=0x0000000102b5b560 ReactiveKit`reabstraction thunk helper <A, B where B: Swift.Error> from @callee_owned (@in ReactiveKit.Event<A, B>) -> (@out ()) to @callee_owned (@in ReactiveKit.Event<A, B>) -> ()partial apply forwarder with unmangled suffix ".48" at Subjects.swift, event=<unavailable>) at Subjects.swift:54
    frame #59: 0x0000000102b4965d ReactiveKit`partial apply for closure #1 in Subject.send(_:) at Subjects.swift:0
    frame #60: 0x0000000102b4c74e ReactiveKit`Subject.forEachObserver(execute=0x0000000102b49600 ReactiveKit`partial apply forwarder for closure #1 ((ReactiveKit.Event<A, B>) -> ()) -> () in ReactiveKit.Subject.send(ReactiveKit.Event<A, B>) -> () at Subjects.swift, self=0x0000606000317a80) at Subjects.swift:81
    frame #61: 0x0000000102b48ef0 ReactiveKit`Subject.send(event=<unavailable>, self=0x0000606000317a80) at Subjects.swift:54
    frame #62: 0x0000000102b487d3 ReactiveKit`Subject.on(event=<unavailable>, self=0x0000606000317a80) at Subjects.swift:50
    frame #63: 0x0000000102b4cc22 ReactiveKit`protocol witness for ObserverProtocol.on(_:) in conformance <A, B> Subject<A, B> at Subjects.swift:0
    frame #64: 0x0000000102c4b187 ReactiveKit`ObserverProtocol.next(element=0x000000010af58b60, self=0x000000010af58b40) at Observer.swift:97
    frame #65: 0x00000001028ae4dd Bond`MutableObservableArray<A>.replace(array=1 value, performDiff=true, self=0x00006040000cd4d0) at ObservableArray.swift:362
```
</details>

Reply:
Ha-ha:) Me too :) But crash occurs only in one place and floats between project builds. I thought that it's sort of compiler issue.

Reply:
@tonyarnold, hmm, is it a Swift object? Can you see if making it a subclass of `NSObject` fixes the issues?

Reply:
They are Swift objects, however, making them subclasses of `NSObject` does not fix this issue.

Reply:
Interestingly, they are Swift objects that conform to a protocol that declares conformance to `AnyObject`. They're definitely not subclasses of `NSObject`, so I'm not sure why they'd even satisfy the conditions of the switch in the protocol proxy.

Reply:
I think I have a fix - give me 20 mins to prepare a PR.

Reply:
@tonyarnold  Can you post a link to your PR?

Reply:
@DivineDominion I'm not sure it was me who ended up fixing this. I think @srdanrasic ended up rewriting some of the protocol proxy to address similar issues to this one. What specifically did you want to see?

Reply:
I'm hunting down a similar crash in another situation (and without Bond), so I am curious what you did to fix this.

Reply:
This issue was very specific to Bond and how it implements protocol proxies by intercepting NSInvocations. The issue was related to Swift -> ObjC type bridging. Here is a first commit from the fix: https://github.com/ReactiveKit/Bond/commit/64f8e95596f11120253ef6006ccfe95fb1a94270

Reply:
Thanks a ton!